src/sentry/api/endpoints/organization_derive_code_mappings.py

```diff
@@ -0,0 +1,116 @@
+
+    def get(self, request: Request, organization: Organization) -> Response:
+        """
+        Get all matches for a filename.
+        ``````````````````
+
+        :pparam string organization_slug: the slug of the organization the
+                                          team should be created for.
+        :param string filename:
+        :auth: required
+        """
+        if not features.has("organizations:derive-code-mappings", organization):
+            return Response(status=status.HTTP_400_BAD_REQUEST)
+
+        filename = request.GET.get("filename")
+        installation, _ = get_installation(organization)
+        if not installation:
+            return self.respond(
+                "Could not find this integration installed on your organization",
+                status=status.HTTP_404_NOT_FOUND,
+            )
+
+        trees = installation.get_trees_for_org()
+        trees_helper = CodeMappingTreesHelper(trees)
+        frame_filename = FrameFilename(filename)
+        possible_code_mappings: List[CodeMappingMatch] = trees_helper.list_file_matches(
+            frame_filename
+        )
+        return Response(serialize(possible_code_mappings), status=status.HTTP_200_OK)
+
+    def post(self, request: Request, organization: Organization) -> Response:
+        """
+        Create a new repository project path config from a filename
+        ``````````````````
+
+        :pparam string organization_slug: the slug of the organization the
+                                          team should be created for.
+        :param int projectId:
+        :param string repoName:
+        :param string defaultBranch:
+        :param string stackRoot:
+        :param string sourceRoot:
+        :auth: required
+        """
+        if not features.has("organizations:derive-code-mappings", organization):
+            return Response(status=status.HTTP_400_BAD_REQUEST)
+
+        installation, organization_integration = get_installation(organization)
+        if not installation:
+            return self.respond(
+                "Could not find this integration installed on your organization",
+                status=status.HTTP_404_NOT_FOUND,
+            )
+
+        project = Project.objects.get(id=request.data.get("projectId"))
+        repo_name = request.data.get("repoName")
+        repository, _ = Repository.objects.get_or_create(
+            name=repo_name,
+            organization_id=organization.id,
+            defaults={
+                "integration_id": organization_integration.integration_id,
+            },
+        )
+
+        stack_root = request.data.get("stackRoot")
+        exists = RepositoryProjectPathConfig.objects.filter(
+            project=project,
+            stack_root=stack_root,
+        )
+        if exists.exists():
+            return self.respond(
+                "Code mapping already exists for this project", status=status.HTTP_400_BAD_REQUEST
```

Comment:
We shouldn't hit this case unless something goes wrong.

Reply:
I would try to add the protection in the object's `create` method below rather than check here.

It might be that updating the code mapping is the right approach (e.g. the customer had the wrong source code root.